src/ast.rs

```diff
@@ -217,24 +217,140 @@ impl Identifier {
+    /// ```rust
+    /// use balena_temen::ast::*;
+    ///
+    /// let identifier: Identifier = "names.wifi".parse().unwrap();
+    /// assert!(identifier.is_canonical());
+    ///
+    /// let identifier: Identifier = "names.wifi[first].id".parse().unwrap();
+    /// assert!(identifier.is_canonical());
+    /// ```
+    ///
+    /// Not canonical identifiers.
+    ///
+    /// ```rust
+    /// use balena_temen::ast::*;
+    ///
+    /// let identifier: Identifier = "names.this".parse().unwrap();
+    /// assert!(!identifier.is_canonical());
+    ///
+    /// let identifier: Identifier = "names[this.index]".parse().unwrap();
+    /// assert!(!identifier.is_canonical());
+    /// ```
+    pub fn is_canonical(&self) -> bool {
+        for v in &self.values {
+            match v {
+                IdentifierValue::This | IdentifierValue::Super => return false,
+                IdentifierValue::Identifier(ref identifier) => {
+                    if !identifier.is_canonical() {
+                        return false;
+                    }
+                }
+                _ => {}
+            };
+        }
+
+        true
+    }
+
+    fn is_relative(&self) -> bool {
         if let Some(first) = self.values.first() {
-            first == &IdentifierValue::This || first == &IdentifierValue::Super
+            match first {
+                IdentifierValue::This | IdentifierValue::Super => true,
+                _ => false,
+            }
         } else {
             false
         }
     }
 
-    /// Check if an identifier is absolute
+    /// Returns the canonical, absolute, identifier with all intermediate
+    /// components normalized and nested identifiers canonicalized.
+    ///
+    /// Nested identifiers (`IdentifierValue::Identifier`) are canonicalized
+    /// too.
+    ///
+    /// # Arguments
+    ///
+    /// * `position` - An identifier position
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use balena_temen::ast::*;
     ///
-    /// An identifier is considered as an absolute one if it does not start with `this`
-    /// or `super` keyword.
-    pub fn is_absolute(&self) -> bool {
-        !self.is_relative()
+    /// let identifier: Identifier = "names".parse().unwrap();
+    /// assert_eq!(identifier.canonicalize(&Identifier::default()).unwrap(), identifier);
+    ///
+    /// let identifier: Identifier = "names.this.id.this.super".parse().unwrap();
+    /// let canonicalized: Identifier = "names".parse().unwrap();
+    /// assert_eq!(identifier.canonicalize(&Identifier::default()).unwrap(), canonicalized);
+    ///
+    /// let identifier: Identifier = "super.id".parse().unwrap();
+    /// let position: Identifier = "wifi[`zrzka`].ssid".parse().unwrap();
+    /// let canonicalized: Identifier = "wifi[`zrzka`].id".parse().unwrap();
+    /// assert_eq!(identifier.canonicalize(&position).unwrap(), canonicalized);
+    /// ```
+    pub fn canonicalize(&self, position: &Identifier) -> Result<Identifier> {
```

Comment:
nit - could benefit from function extraction, to get `position_values`, could even be a private method on `self`

Reply:
Fixed, 2nd method introduced to make it smaller.